common/src/Model/Brush.cpp

```diff
@@ -969,20 +1007,26 @@ namespace TrenchBroom {
                 swap(remaining, moving);
                 delta = -delta;
             }
-            
-            // Now check if any of the moving vertices would travel into or through the remaining fragment.
+
+            // Now check if any of the moving vertices would travel through the remaining fragment.
             for (const BrushVertex* vertex : moving.vertices()) {
                 const Vec3& oldPos = vertex->position();
                 const Vec3 newPos = oldPos + delta;
                 
+                // Skip moving vertices that end up inside the remaining fragment.
+                // These are handled in the allowVertexRemoval section above.
+                if (remaining.contains(newPos))
+                    continue;
+                
                 for (const BrushFaceGeometry* face : remaining.faces()) {
                     if (face->pointStatus(oldPos) == Math::PointStatus::PSAbove &&
```

Comment:
We still need the raytracing, e.g. moving the top face to the right, it will cross the plane of the highlighted face on the right side of the lower cube:
<img width="336" alt="screen shot 2016-12-27 at 1 20 03 pm" src="https://cloud.githubusercontent.com/assets/239161/21507741/4d10a858-cc38-11e6-98f0-9781391ae018.png">

The current code is checking for the face where the ray enters the polyhedron; checking for the exit point should work as well and may be clearer